### ISSUE TITLE
Upgrade AWS SDK to v2

### ISSUE THREAD

**Issue by sumerjabri** [craftercms/craftercms#6617]
### Latest version

- [X] The issue is in the develop branch
- [X] The issue is in the latest released 4.1.x
- [X] The issue is in the latest released 4.0.x
- [X] The issue is in the latest released 3.1.x

### Duplicates

- [X] I have searched the existing issues

### Is your enhancement request related to a problem? Please describe.

AWS is ending support for AWS Java SDK v1: https://aws.amazon.com/blogs/developer/announcing-end-of-support-for-aws-sdk-for-java-v1-x-on-december-31-2025/

### Describe the solution you'd like

Upgrade all modules to use AWS Java SDK v2

https://docs.aws.amazon.com/sdk-for-java/latest/developer-guide/home.html

**Comment by sumerjabri** [comment]
CC @jmendeza 

Note that this may provide better/faster S3 operations, and perhaps bulk ops.

**Comment by Euquimides** [comment]
Hi @jmendeza and  @phuongnq ! While validating this ticket I noticed that when uploading an image to S3 through the data source `Image Uploaded to S3 Repository`, if the filename of the image has spaces in it, such as `Screenshot from 2022-03-30 08-31-25.png`, the upload image will not load inside the form, since spaces seems to be encoded different in AWS vs Studio. 

For example, in S3 the resource path will be:

`awscraftercms/images/Screenshot+from+2024-05-29+16-04-37.png`

While in Studio, the following error will show up in the logs:

```
[ERROR] 2024-07-09T16:51:38,907 [http-nio-8080-exec-5] [test3] [impl.DefaultExceptionHandler] | GET http://localhost:8080/remote-assets/s3/crafter-test/awscraftercms/images/Screenshot%20from%202024-06-11%2012-42-40.png failed 
java.io.FileNotFoundException: S3Resource{profile=S3Profile{profileId='crafter-test', region='us-east-1', endpoint='null', bucketName='awscraftercms', prefix='null', pathStyleAccessEnabled=false}, key='awscraftercms/images/Screenshot%20from%202024-06-11%2012-42-40.png'} not found
```
![image](https://github.com/craftercms/craftercms/assets/90878288/6f28a5c1-ee58-4903-a937-73fe03f5eabe)

Could you please check this out? Thanks!

**Comment by jmendeza** [comment]
> Hi @jmendeza and @phuongnq ! While validating this ticket I noticed that when uploading an image to S3 through the data source `Image Uploaded to S3 Repository`, if the filename of the image has spaces in it, such as `Screenshot from 2022-03-30 08-31-25.png`, the upload image will not load inside the form, since spaces seems to be encoded different in AWS vs Studio.
> 
> For example, in S3 the resource path will be:
> 
> `awscraftercms/images/Screenshot+from+2024-05-29+16-04-37.png`
> 
> While in Studio, the following error will show up in the logs:
> 
> ```
> [ERROR] 2024-07-09T16:51:38,907 [http-nio-8080-exec-5] [test3] [impl.DefaultExceptionHandler] | GET http://localhost:8080/remote-assets/s3/crafter-test/awscraftercms/images/Screenshot%20from%202024-06-11%2012-42-40.png failed 
> java.io.FileNotFoundException: S3Resource{profile=S3Profile{profileId='crafter-test', region='us-east-1', endpoint='null', bucketName='awscraftercms', prefix='null', pathStyleAccessEnabled=false}, key='awscraftercms/images/Screenshot%20from%202024-06-11%2012-42-40.png'} not found
> ```
> 
> ![image](https://private-user-images.githubusercontent.com/90878288/347172907-6f28a5c1-ee58-4903-a937-73fe03f5eabe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1Njc1MjEsIm5iZiI6MTcyMDU2NzIyMSwicGF0aCI6Ii85MDg3ODI4OC8zNDcxNzI5MDctNmYyOGE1YzEtZWU1OC00OTAzLWE5MzctNzNmZTAzZjVlYWJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA5VDIzMjAyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyMGIyNTk3MDVmMmE1ZDUzYzE5ZWZkNTdlOTAxOWEwNDliODk2ZjRjNDI4OWQ1MGI5ZDkwZWY1MWU1NGJhY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JsBwjl66ovhDNhuGyk2TIRU-JAC_MBAvNJiGncRKhBI)
> 
> Could you please check this out? Thanks!

@Euquimides would you please confirm if this is behaviour is specific to 4.2? Otherwise it might not be directly related to this upgrade

**Comment by phuongnq** [comment]
The problem with space is that the SDK API HeadObject (https://docs.aws.amazon.com/AmazonS3/latest/API/API_HeadObject.html) API does not support getting an S3 Object with an encoded key. 

We can decode the S3 key before retrieving it such as `Screenshot%20from%202024-06-11%2012-42-40.png` => `Screenshot from 2024-06-11 12-42-40.png` but then we don't know which key is encoded and which is not.
For example, the file with the name `Screenshot%20from%202024-06-11%2012-42-40.png` is a valid key in S3 and you can directly upload it to S3. In this case, decoding the key before retrieving it via the SDK is not a correct move. Some sample files in my S3 bucket:

![2024-07-10_17-29](https://github.com/craftercms/craftercms/assets/2996543/39000bf5-bc2b-4e36-8d31-f0ebee620555)

Maybe it would be a good convention that we don't accept files with space while working with S3? @jmendeza @sumerjabri

**Comment by Euquimides** [comment]
> > Hi @jmendeza and @phuongnq ! While validating this ticket I noticed that when uploading an image to S3 through the data source `Image Uploaded to S3 Repository`, if the filename of the image has spaces in it, such as `Screenshot from 2022-03-30 08-31-25.png`, the upload image will not load inside the form, since spaces seems to be encoded different in AWS vs Studio.
> > For example, in S3 the resource path will be:
> > `awscraftercms/images/Screenshot+from+2024-05-29+16-04-37.png`
> > While in Studio, the following error will show up in the logs:
> > ```
> > [ERROR] 2024-07-09T16:51:38,907 [http-nio-8080-exec-5] [test3] [impl.DefaultExceptionHandler] | GET http://localhost:8080/remote-assets/s3/crafter-test/awscraftercms/images/Screenshot%20from%202024-06-11%2012-42-40.png failed 
> > java.io.FileNotFoundException: S3Resource{profile=S3Profile{profileId='crafter-test', region='us-east-1', endpoint='null', bucketName='awscraftercms', prefix='null', pathStyleAccessEnabled=false}, key='awscraftercms/images/Screenshot%20from%202024-06-11%2012-42-40.png'} not found
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ![image](https://private-user-images.githubusercontent.com/90878288/347172907-6f28a5c1-ee58-4903-a937-73fe03f5eabe.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjA1Njc1MjEsIm5iZiI6MTcyMDU2NzIyMSwicGF0aCI6Ii85MDg3ODI4OC8zNDcxNzI5MDctNmYyOGE1YzEtZWU1OC00OTAzLWE5MzctNzNmZTAzZjVlYWJlLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA3MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNzA5VDIzMjAyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQyMGIyNTk3MDVmMmE1ZDUzYzE5ZWZkNTdlOTAxOWEwNDliODk2ZjRjNDI4OWQ1MGI5ZDkwZWY1MWU1NGJhY2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.JsBwjl66ovhDNhuGyk2TIRU-JAC_MBAvNJiGncRKhBI)
> > Could you please check this out? Thanks!
> 
> @Euquimides would you please confirm if this is behaviour is specific to 4.2? Otherwise it might not be directly related to this upgrade

Hi @jmendeza ! I just checked this out in v4.1.6, seems it is only happening in v4.2.

**Comment by tfield** [comment]
> Maybe it would be a good convention that we don't accept files with space while working with S3?

That would be an unfortunate regression. Users like to upload docs as they have them normally, and spaces in filenames are the norm. It would be preferable to document the encoding requirements and let the developers manage encoding URLs, file operations, etc.

**Comment by phuongnq** [comment]
I sent PR to decode the key before retrieving it from S3 to match the behavior in 4.1.x.
For the context, this is not an issue with the SDK itself, but a change in Spring 6 `HandlerMapping.PATH_WITHIN_HANDLER_MAPPING_ATTRIBUTE`.

In Spring 5: this path value is not encoded but in Spring 6, it is encoded for security reasons. We should only decode this when needed. In this case, this is when we retrieve the S3 assets.

**Comment by Euquimides** [comment]
Hi @sumerjabri and @rart ! While validating this ticket I noticed an issue in the UI that doesn't happen in 4.1.x but does in 4.2. Maybe the AWS SDK update might had an impact? The reference ticket is here: https://github.com/craftercms/craftercms/issues/7032

**Comment by Euquimides** [comment]
Hi! Enhancement implemented in Crafter v4.2. Thank you! Closing this ticket.
